Adds documentation for half_t and bhalf_t (#702)
kokkos/kokkos-core-wiki: docs/source/API/core/Half-precision-types.rst

.. _api-Half-precision-types:

Half precision types
====================

.. warning::

>    ``half_t`` and ``bhalf_t`` are still in the namespace ``Kokkos::Experimental``

Reviewer: ```suggestion
   ``half_t`` and ``bhalf_t`` are still in the ``Kokkos::Experimental`` namespace.
```

Types
-----
Kokkos offers portable half precision types under the name ``Kokkos::Experimental::half_t`` and ``Kokkos::Experimental::bhalf_t``.
 - ``half_t`` represents the standard half precision float, with a 1 bit Sign Bit, 5 bits Exponent and 10 bits Significand.
 - ``bhalf_t`` corresponds to the type known as 'brain half' and has a 1 bit Sign Bit, 8 bits Exponent and 7 bits Significand.

This types will either map to the current backend own types (for instance ``__half`` on Cuda), or to ``float`` if no such type is available.

The macros ``KOKKOS_HALF_T_IS_FLOAT`` and ``KOKKOS_BHALF_T_IS_FLOAT`` are set to ``true`` when ``half_t`` and ``bhalf_t`` are mapped to ``float``, and ``false`` when mapped to its own separate type.

Functions
---------
The following table list the standard mathematical functions that can be used with the ``half_t`` and ``bhalf_t`` type.

> In addition, for the Cuda and Sycl backends, the marked functions are performed using specific half precision functions and may thus be quicker.

Reviewer: ```suggestion
In addition, for the Cuda and SYCL backends, the marked functions are performed using specific half precision functions and may thus be more performant.
```

The default behaviour if the specific function doesn't exist is to cast the half precision float to ``float``, perform the operation with the standard function and cast back the result to half precision.

.. csv-table::

>    :header: "Function", "``half_t`` Cuda", "``bhalf_t`` Cuda", "``half_t`` Sycl", "``bhalf_t`` Sycl"

Reviewer: ```suggestion
   :header: "Function", "``half_t`` Cuda", "``bhalf_t`` Cuda", "``half_t`` SYCL", "``bhalf_t`` SYCL"
```

   "fabs","X", "X", "X", "X"
   "fmod", , , "X", 
   "remainder", , , "X", 
   "fmax","X¹", "X¹", "X", "X"
   "fmin","X¹", "X¹", "X", "X"
   "fdim", "X", "X", "X", 
   "exp", "X", "X", "X", "X"
   "exp2", "X", "X", "X", "X"
   "expm1", , , "X", 
   "log", "X", "X", "X", "X"
   "log10", "X", "X", "X", "X"
   "log2", "X", "X", "X", "X"
   "log1p", , , "X", 
   "pow", , , "X", 
   "sqrt", "X", "X", "X", "X"
   "cbrt", , , "X", 
   "hypot", , , "X", 
   "sin", "X", "X", "X", "X"
   "cos", "X", "X", "X", "X"
   "tan", , , "X", 
   "asin", , , "X", 
   "acos", , , "X", 
   "atan", , , "X", 
   "atan2", , , "X", 
   "sinh", , , "X", 
   "cosh", , , "X", 
   "tanh", , , "X", 
   "asinh", , , "X", 
   "acosh", , , "X", 
   "atanh", , , "X", 
   "erf", , , "X", 
   "erfc", , , "X", 
   "tgamma", , , "X", 
   "lgamma", , , "X", 
   "ceil", "X", "X", "X", "X"
   "floor", "X", "X", "X", "X"
   "trunc", "X", "X", "X", "X"
   "round", , , "X", 
   "nearbyint", "X", "X", , 
   "logb", , , "X", 
   "nextafter", , , "X", 
   "copysign", , , "X", 
   "isfinite", , , "X", 
   "isinf", "X²", "X²", "X", 
   "isnan", "X", "X", "X", "X"
   "signbit", , , "X", 

¹Only if GPU_ARCH >= 80

> ²Not for nvcc version between 12.1 and 13.0 due to a compiler bug

Author: I noticed the issue when doing the improvement for the SYCL and Cuda backends, but not tested with enough different version, turned out it affected a lot less version than I thought:

https://docs.nvidia.com/cuda/archive/12.3.2/cuda-toolkit-release-notes/index.html

> Resolved Issues
> 
>     The __hisinf() Math APIs from cuda_fp16.h and cuda_bf16.h headers were silently producing wrong results if compiled with the -std=c++20 compiler option because of an underlying nvcc compiler issue, resolved in version 12.3.

I will make a bugfix in Kokkos too.

> ²Not for nvcc version between 12.1 and 13.0 due to a compiler bug

Reviewer 2: could you link the bugfix pr here, please?

> ²Not for nvcc version between 12.1 and 13.0 due to a compiler bug

Author: Sorry, I forgot to do it, it is this one : https://github.com/kokkos/kokkos/pull/8290.
